tests/client/core/test_middleware.js

```diff
@@ -248,6 +255,50 @@ describe('Trailing Slashes Middleware', () => {
     assert.notOk(fakeNext.called);
   });
 
+  it('should not add trailing slashes if the URL has an exception', () => {
+    const fakeReq = {
+      originalUrl: '/no/trailing',
+      headers: {},
+    };
+    trailingSlashesMiddleware(
+      fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.notOk(fakeRes.redirect.called);
+    assert.ok(fakeNext.called);
+  });
+
+  it('should handle trailing slash exceptions with $lang', () => {
+    const fakeReq = {
+      originalUrl: '/en-US/no/trailing',
+      headers: {},
+    };
+    trailingSlashesMiddleware(
+      fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.notOk(fakeRes.redirect.called);
+    assert.ok(fakeNext.called);
+  });
+
+  it('should handle trailing slash exceptions with $clientApp', () => {
+    const fakeReq = {
+      originalUrl: '/firefox/no/trailing',
+      headers: {},
+    };
+    trailingSlashesMiddleware(
+      fakeReq, fakeRes, fakeNext, { _config: fakeConfig });
+    assert.notOk(fakeRes.redirect.called);
+    assert.ok(fakeNext.called);
+  });
+
+  it('should handle trailing slash exceptions with $lang/$clientApp', () => {
```

Comment:
I may have missed it, but it would be good to ensure that clientApp and or locale prefixed urls still get redirs if not excluded.